lib/extensions/attachable/active_record/base.rb

```diff
@@ -2,31 +2,102 @@
+    #   the column type should be string or text.
+    #   This can be a symbol or an array of symbols.
+    # @example Has many attachments on a column
+    #   has_many_attachments on: :description #=> description is associated with the attachments
+    #   of the model, updating description will result in attachments changing.
+    def has_many_attachments(options = {}) # rubocop:disable Style/PredicateName
+      include HasManyAttachments
+
+      class_attribute :attachable_columns
+      if options[:on]
+        self.attachable_columns = Array(options[:on])
+        before_save :update_attachment_references
+      end
+    end
+
+    def has_one_attachment # rubocop:disable Style/PredicateName
+      include HasOneAttachment
+    end
+  end
+
+  module HasManyAttachments
+    extend ActiveSupport::Concern
+
+    included do
       has_many :attachment_references, as: :attachable, class_name: "::#{AttachmentReference.name}",
                                        inverse_of: :attachable, dependent: :destroy, autosave: true
       # Attachment references can substitute attachments, so allow access using the `attachments`
       # identifier.
       alias_method :attachments, :attachment_references
+    end
+
+    def files=(files)
+      files.each do |file|
+        attachment_references.build(file: file)
+      end
+    end
+
+    private
 
-      define_method(:files=) do |files|
-        files.each do |file|
-          attachment_references.build(file: file)
+    # Delete the attachment references which are not in the content.
+    def update_attachment_references
+      return if attachment_references.empty?
+
+      attachment_reference_ids = active_attachment_reference_ids
+      attachment_references.each do |attachment_reference|
+        unless attachment_reference_ids.include?(attachment_reference.id)
+          attachment_reference.mark_for_destruction
         end
       end
     end
 
-    def has_one_attachment # rubocop:disable Style/PredicateName
-      include SingularInstanceMethods
+    # Find all attachment_reference ids in the columns specified.
+    #
+    # @return [Array<Integer>]
+    def active_attachment_reference_ids
+      attachment_reference_ids = []
+      columns = self.class.attachable_columns || []
+      columns.each do |column|
+        content = send(column)
+        attachment_reference_ids += parse_attachment_reference_ids_from_content(content)
+      end
+
+      attachment_reference_ids
+    end
+
+    def parse_attachment_reference_ids_from_content(content)
+      ids = []
+      doc = Nokogiri::HTML(content)
+      doc.css('img').each do |image|
+        id = parse_attachment_reference_id_from_url(image['src'])
+        ids << id if id
+      end
 
+      ids
+    end
+
+    ATTACHMENT_ID_REG = /\/attachments\/(\d+)$/
+    def parse_attachment_reference_id_from_url(url)
+      result = url.match(ATTACHMENT_ID_REG)
```

Comment:
Haven't considered this case yet, since we have multi tenancy, the valid domains can be very dynamic. Think this need further discussion, I added a TODO.